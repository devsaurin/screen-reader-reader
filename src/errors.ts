export class SRRError extends Error {
  public details: Array<unknown>;

  constructor(public code: ErrorCodes, ...details: Array<unknown>) {
    super();

    this.details = details;
  }
}

export function logError(err: SRRError): void {
  process.stderr.write(ERROR_MESSAGES[err.code](...err.details));
}

const FATAL_ERROR = `
If you see this, something is really wrong!
Please create an issue at https://github.com/phenomnomnominal/screen-reader-reader! ❤️
`;

const CONFIG_ERROR = `
If you see this, you may have an issue with your config!
`;

const TEST_ERROR = `
If you see this, you may have an issue with your setup!
`;

type ErrorMessageFactory = (...details: Array<unknown>) => string;
const ERROR_MESSAGES: Record<ErrorCodes, ErrorMessageFactory> = {
  [ErrorCodes.NOT_IMPLEMENTED]: (methodName: unknown): string => `
${FATAL_ERROR}
  Oops, "${methodName}" is not implemented!

`,

  [ErrorCodes.INVALID_SCREENREADER_NAME]: (
    name: unknown,
    possibleNames: unknown
  ): string => `
${CONFIG_ERROR}
"${name}" is not a valid value for \`screenreader\`!
Valid values for \`screenreader\` are:

  "${(possibleNames as Array<string>).join(', ')}"
    
`,
  [ErrorCodes.INVALID_POLL_TIMEOUT]: (pollTimeout: unknown): string => `
${CONFIG_ERROR}
"${pollTimeout}" is not a valid value for \`pollTimeout\`!
Valid values for \`pollTimeout\` are:

  Any number greater than \`100\`.

`,
  [ErrorCodes.INVALID_STABLE_TIMEOUT]: (stableTimeout: unknown): string => `  
${CONFIG_ERROR}
"${stableTimeout}" is not a valid value for \`stableTimeout\`!
Valid values for \`stableTimeout\` are:

  Any number greater than \`0\`.

`,
  [ErrorCodes.INVALID_WAIT_FOR_STABLE]: (waitForStable: unknown): string => `
${CONFIG_ERROR}
"${waitForStable}" is not a valid value for \`waitForStable\`!
Valid values for \`waitForStable\` are:

  \`true\` or \`false\`

`,
  [ErrorCodes.INVALID_FILTERS]: (filters: unknown): string => `
${CONFIG_ERROR}
  "${filters}" is not a valid value for \`filters\`!
  \`filters\` must be an Array.

`,
  [ErrorCodes.INVALID_FILTER]: (filter: unknown): string => `
${CONFIG_ERROR}
  "${filter}" is not a valid filter!
  Each filter must be a RegExp.

`,
  [ErrorCodes.INVALID_MAPPERS]: (mappers: unknown): string => `
${CONFIG_ERROR}
  "${mappers}" is not a valid value for \`mappers\`!
  \`mappers\` must be an Array.
 
`,
  [ErrorCodes.INVALID_MAPPER]: (mapper: unknown): string => `
${CONFIG_ERROR}
"${mapper}" is not a valid mapper!
Each mapper must be with this type signature:

  (str: string) => string

`,

  [ErrorCodes.NO_SCREENREADER_OUTPUT]: (): string => `
${TEST_ERROR}
There was no recorded screen-reader output!

This could mean:

  ✨There really was no output - was that intentional?
  ✨Something isn't working correctly - does the screen-reader show output when run manually?

`,
  [ErrorCodes.CANT_DETECT_SCREENREADER]: (): string => `
${TEST_ERROR}
Couldn't automatically detect which screen-reader to start!

screen-reader-reader tries to automatically choose a screen-reader for you, based on the following:

  ⭐️ "VoiceOver" - if you are running MacOS.
  ⭐️ "NVDA" - if you are running Windows with NVDA installed in the default directory.

`,
  [ErrorCodes.CANT_START_WHILE_STARTED]: (): string => `
${TEST_ERROR}
You can't start screen-reader-reader again while it is already running!

`
};

export const enum ErrorCodes {
  NOT_IMPLEMENTED,

  INVALID_SCREENREADER_NAME,
  INVALID_POLL_TIMEOUT,
  INVALID_STABLE_TIMEOUT,
  INVALID_WAIT_FOR_STABLE,
  INVALID_FILTERS,
  INVALID_FILTER,
  INVALID_MAPPERS,
  INVALID_MAPPER,

  NO_SCREENREADER_OUTPUT,
  CANT_DETECT_SCREENREADER,
  CANT_START_WHILE_STARTED
}
