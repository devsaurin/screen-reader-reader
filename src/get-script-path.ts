import * as path from 'path';

export function getScriptPath(scriptName: string): string {
  return path.resolve(__dirname, `../scripts/${scriptName}`);
}
