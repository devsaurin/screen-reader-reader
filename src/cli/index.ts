import minimist from 'minimist';

import { logError } from '../errors';

import { setup } from './setup';
import { start } from './start';

export async function cli(argv: Array<string>): Promise<void> {
  const [first, ...rest] = argv;

  try {
    if (first === 'setup') {
      await setup(minimist(rest));
      return;
    }
    await start(minimist(argv));
    return;
  } catch (e) {
    logError(e);
  }
}
