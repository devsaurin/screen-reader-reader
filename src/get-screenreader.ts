import * as path from 'path';
import { existsSync } from 'fs';

import { NVDA } from './screenreaders/nvda';
import { VoiceOver } from './screenreaders/voiceover';
import { Options } from './options';

export type ScreenReader = NVDA | VoiceOver;

export const enum ScreenReaderName {
  voiceover,
  nvda
}

const DEFAULT_NDVA_PATH = path.join(
  'C:/',
  'Program Files (x86)',
  'NVDA',
  'nvda.exe'
);

export type ScreenReaderConstructor = (options: Options) => ScreenReader;

const CONSTRUCTORS: Record<ScreenReaderName, ScreenReaderConstructor> = {
  [ScreenReaderName.nvda]: (options: Options) => new NVDA(options),
  [ScreenReaderName.voiceover]: (options: Options) => new VoiceOver(options)
};

export function getScreenReaderName(
  screenReader?: string | ScreenReaderName
): ScreenReaderName {
  if (typeof screenReader === 'string') {
    if (screenReader.toLowerCase() === 'nvda') {
      return ScreenReaderName.nvda;
    }
    if (screenReader.toLowerCase() === 'voiceover') {
      return ScreenReaderName.voiceover;
    }

    throw new Error(`
"${screenReader}" is not a valid screen reader name. Valid options are:

  * "VoiceOver"
  * "NVDA"
    `);
  }

  if (screenReader) {
    return screenReader;
  }

  if (process.platform === 'darwin') {
    return ScreenReaderName.voiceover;
  }
  if (existsSync(DEFAULT_NDVA_PATH)) {
    return ScreenReaderName.nvda;
  }

  throw new Error(`
Could not automatically detect screen reader.
Default for MacOS is "VoiceOver".
Default for Windows is "NVDA" which must installed at "${DEFAULT_NDVA_PATH}".
  `);
}

export function instantiateScreenReader(options: Options): ScreenReader {
  return CONSTRUCTORS[options.screenreader](options);
}
