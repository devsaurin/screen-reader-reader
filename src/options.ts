import { ScreenReaderName, getScreenReaderName } from './get-screenreader';

export type Options = {
  screenreader: ScreenReaderName;
  pollTimeout: number;
  stableTimeout: number;
  waitForStable: boolean;
  filters: Array<RegExp>;
  mappers: Array<(t: string) => string>;
};

const DEFAULT_POLL_TIMEOUT = 100;
const DEFAULT_STABLE_TIMEOUT = 10000;

export function getOptions(options: Partial<Options>): Options {
  return {
    screenreader: getScreenReaderName(options.screenreader),
    pollTimeout:
      options.pollTimeout != null ? options.pollTimeout : DEFAULT_POLL_TIMEOUT,
    stableTimeout:
      options.stableTimeout != null
        ? options.stableTimeout
        : DEFAULT_STABLE_TIMEOUT,
    waitForStable: options.waitForStable || false,
    filters: options.filters || [],
    mappers: options.mappers || []
  };
}
