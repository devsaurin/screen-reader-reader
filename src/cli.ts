import minimist from 'minimist';

import { startScreenReader } from './index';
import { Options } from './options';
import { getScreenReaderName } from './get-screenreader';

type CLIOptions = {
  screenreader: string;
  filters: Array<string>;
  pollTimeout: number;
  stableTimeout: number;
};

export async function cli(argv: Array<string>): Promise<void> {
  const options = mapCLIArgs(minimist(argv) as Partial<CLIOptions>);

  const result = await startScreenReader(options);

  process.stdout.write(result);
}

function mapCLIArgs(
  cliOptions: Partial<CLIOptions>
): Partial<Options> & { waitForStable: true } {
  return {
    waitForStable: true,
    screenreader: getScreenReaderName(cliOptions.screenreader),
    filters: (cliOptions.filters || []).map(f => new RegExp(f)),
    pollTimeout: cliOptions.pollTimeout,
    stableTimeout: cliOptions.stableTimeout
  };
}
