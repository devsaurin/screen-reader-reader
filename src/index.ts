import { instantiateScreenReader } from './get-screenreader';
import { Options, getOptions } from './options';

let reading = false;

export async function startScreenReader(
  options: Partial<Options> & { waitForStable: true }
): Promise<string>;

export async function startScreenReader(
  options: Partial<Options> & { waitForStable?: false }
): Promise<() => Promise<string>>;

export async function startScreenReader(
  options: Partial<Options>
): Promise<string | (() => Promise<string>)> {
  if (reading) {
    throw new Error(
      'You can only have one screen reader reader running at a time'
    );
  }

  reading = true;

  const finalOptions = getOptions(options);

  const screenreader = instantiateScreenReader(finalOptions);

  const stop = await screenreader.start();

  if (finalOptions.waitForStable) {
    return await stopScreenReader(finalOptions.waitForStable);
  }

  return function(): Promise<string> {
    return stopScreenReader();
  };

  async function stopScreenReader(waitForStable?: boolean): Promise<string> {
    try {
      return await stop(waitForStable);
    } finally {
      reading = false;
    }
  }
}
