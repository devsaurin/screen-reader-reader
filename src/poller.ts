import { Options } from './options';

export type UpdateStable = () => void;
export type Work = (updateStable: UpdateStable) => void;
export type Stop = (waitForStable?: boolean) => Promise<void>;

export function poll(options: Options, work: Work): Stop {
  const { stableTimeout, pollTimeout } = options;

  const stop = createStop();

  let keepPolling: NodeJS.Timeout | null = null;
  updateStableTimeout();

  const polling = runPoll();

  return stop;

  function createStop(): Stop {
    return async (waitForStable = false): Promise<void> => {
      if (waitForStable !== true) {
        keepPolling = null;
      }
      await polling;
    };
  }

  async function runPoll(): Promise<void> {
    while (keepPolling !== null) {
      work(updateStableTimeout);
      await delay(pollTimeout);
    }
  }

  function updateStableTimeout(): void {
    if (keepPolling) {
      clearTimeout(keepPolling);
    }
    keepPolling = setTimeout(() => {
      keepPolling = null;
    }, stableTimeout);
  }
}

function delay(time: number): Promise<void> {
  return new Promise(resolve => {
    setTimeout(resolve, time);
  });
}
