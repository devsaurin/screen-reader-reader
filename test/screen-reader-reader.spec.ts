import { startScreenReader } from '../src/index';

const REALLY_BIG_TIMEOUT = 1000000;

describe('screen-reader-reader', () => {
  [
    { url: 'https://www.google.com' },
    {
      url: 'https://www.typescriptlang.org/',
      stableTimeout: 20000,
      filters: [/!/],
      mappers: [(t: string): string => t.replace(/\d{5,}/g, 'REPLACED')]
    },
    { url: 'https://www.apple.com/voiceover/info/guide/_1121.html' },
    {
      url: 'https://jestjs.io/en/help',
      filters: [/link \d+ stargazers on Github/]
    }
  ].forEach(({ url, stableTimeout, filters }) => {
    it(`should get the screen reader output for some web pages: ${url}`, async () => {
      jest.setTimeout(REALLY_BIG_TIMEOUT);

      await page.goto(url);

      const result = await startScreenReader({
        waitForStable: true,
        stableTimeout,
        filters
      });

      expect(result).toMatchSnapshot();
    });
  });
});
