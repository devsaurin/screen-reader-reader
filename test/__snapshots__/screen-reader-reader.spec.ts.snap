// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`screen-reader-reader should get the screen reader output for some web pages: https://jestjs.io/en/help 1`] = `"banner link JEST list  7 items link Docs link API visited link Help link Blog link English Menu pop-up combo box Search link GitHub heading level 2 Need help? heading level 2 Browse the docs list  4 items ‚Ä¢ Learn how to link get started ‚Ä¢  bullet link Troubleshoot ‚Ä¢ Learn how to link configure Jest ‚Ä¢ Look at the full link API Reference heading level 2 Join the community list  2 items ‚Ä¢ Join the link #jest channel on link Reactiflux link existing questions tagged with jestjs or link ask your own heading level 2 Stay up to date list  3 items ‚Ä¢ Follow link Jest ‚Ä¢ Subscribe to the link Jest blog ‚Ä¢ Look at the link changelog Jest group Jest image heading level 5 Docs link Getting Started link Guides link API Reference heading level 5 Community link Stack Overflow link Jest Chat link Twitter heading level 5 More link Blog link GitHub link Star facebook/jest on GitHub link image Facebook Open Source"`;

exports[`screen-reader-reader should get the screen reader output for some web pages: https://www.apple.com/voiceover/info/guide/_1121.html 1`] = `"heading level 2 About VoiceOver heading level 2 About the VoiceOver cursor, mouse pointer, and keyboard focus heading level 2 About keyboard shortcuts and VoiceOver heading level 2 Using function keys on some keyboards To change the default behavior of function keys: list  2 items"`;

exports[`screen-reader-reader should get the screen reader output for some web pages: https://www.google.com 1`] = `"link Gmail link Images Google apps collapsed button link Sign in Search combo box Search by voice button Google Search button I'm Feeling Lucky button Google image Google offered in: link svenska A privacy reminder from Google REMIND ME LATER button REMIND ME LATER REVIEW NOW button Sweden link Privacy link Terms Settings pop-up button link Advertising link Business link About link How Search works"`;

exports[`screen-reader-reader should get the screen reader output for some web pages: https://www.typescriptlang.org/ 1`] = `"link Skip to main content link image TypeScript Home Page list  4 items link Documentation link Download link Connect link Playground link Download the latest version of TypeScript the latest version today! link Learn more about Microsoft's Cookie Policy Logo image link Jump to Downloads link Documentation heading level 2 link #iHeartTypeScript Previous button Dave Herman @littlecalculist I ported my first nontrivial JS lib to link @typescriptlang Gabriela Mendes  image Gabriela Mendes @Kappyh link @typescriptlang is really awesome O-O' Thiago Script üå¥üòéüèñ  image Thiago Script üå¥üòéüèñ @thiagoviski link #TypeScript is really awesome! I'm glad to see people are using it in some of link #Preact link #FrontEnd link #Webpack link #Programming Next button heading level 3 2 items Starts and ends with JavaScript JavaScript heading level 3 2 items Strong tools for large apps Strong tools for large apps heading level 3 2 items State of the art JavaScript State of the art JavaScript link Learn it fast heading level 1 Get TypeScript definition list  2 items INSTALL npm install -g typescript definition list  2 items COMPILE heading level 3 Visual Studio list  3 items link Visual Studio 2019 link Visual Studio Code link Visual Studio 2017 list  3 items link Sublime Text link Atom link Eclipse list  3 items link Emacs link WebStorm link Vim heading level 1 See TypeScript in Action heading level 1 Open Source list  3 items ‚ñ†  bullet Read link the specification on GitHub or download it as a link docx or link pdf ‚ñ†  bullet Play with the link bits and link file bugs ‚ñ†  bullet Join the link #typescript Twitter discussion and link follow the GitHub project link Get the source code heading level 1 Why TypeScript? Previous button Next button Next link See more friends of TypeScript Contribute Help shape the future of TypeScript by joining our community today and send us pull requests via our link GitHub repository Learn link Docs - To help you get started link GitHub & link Stack Overflow - File and discuss issues link Blog & link Twitter - Stay up to date Made with in Redmond link Follow @typescript Microsoft logo image ¬©2012-2019 Microsoft link Privacy"`;
